packages/pipeline-editor/src/PipelineEditorWidget.tsx

```diff
@@ -349,44 +354,87 @@ const PipelineWrapper: React.FC<IProps> = ({
             </p>
           ),
           buttons: [Dialog.cancelButton(), Dialog.okButton()]
-        }).then(result => {
+        }).then(async result => {
           isDialogAlreadyShowing.current = false;
           if (result.button.accept) {
             // proceed with migration
             console.log('migrating pipeline');
-            const migratedPipeline = migrate(
-              contextRef.current.model.toJSON(),
-              pipeline => {
-                // function for updating to relative paths in v2
-                // uses location of filename as expected in v1
-                for (const node of pipeline.nodes) {
-                  node.app_data.filename = PipelineService.getPipelineRelativeNodePath(
-                    contextRef.current.path,
-                    node.app_data.filename
-                  );
-                }
-                return pipeline;
+            let migrationPalette = palette;
+            const pipelineJSON: any = contextRef.current.model.toJSON();
+            const oldRuntime = pipelineJSON?.pipelines[0].app_data.runtime;
+            if (oldRuntime === 'kfp' || oldRuntime === 'airflow') {
+              migrationPalette = await componentFetcher(oldRuntime);
+            }
+            try {
+              const migratedPipeline = migrate(
+                pipelineJSON,
+                pipeline => {
+                  // function for updating to relative paths in v2
+                  // uses location of filename as expected in v1
+                  for (const node of pipeline.nodes) {
+                    node.app_data.filename = PipelineService.getPipelineRelativeNodePath(
+                      contextRef.current.path,
+                      node.app_data.filename
+                    );
+                  }
+                  return pipeline;
+                },
+                migrationPalette
+              );
+              contextRef.current.model.fromString(
+                JSON.stringify(migratedPipeline, null, 2)
```

Comment:
Why are you converting this to a string before saving it? Couldn't you just use `model.fromJSON`?

Reply:
This step prettifies the json